syntax = "proto2";
// Proto description of the gif format.
// Refer to: https://www.fileformat.info/format/gif/egff.htm
// https://www.matthewflickinger.com/lab/whatsinagif/bits_and_bytes.asp

// TODO: Desribe messages included as empty stubs

message LogicalScreenDescriptor {
    required uint32 ScreenWidth = 1;
    required uint32 ScreenHeight= 2;
    required uint32 Packed = 3;
    required uint32 BackgroundColor = 4;
    required uint32 AspectRatio = 5;
}

message GlobalColorTable {
    required bytes colors = 1;
}

message ImageDescriptor {
  required uint32 Seperator =1;
  required uint32 Left =2;
  required uint32 Top = 3;
  required uint32 Width = 4;
  required uint32 Height = 5;
  required uint32 Packed = 6;
}

message ImageChunk {
  oneof chunk_oneof {
    BasicChunk basic = 1;
    PlainTextChunk plaintext = 2;
    ApplicationExtension appExt = 3;
    CommentExtension comExt = 4;
  }
}

message GraphicControlExtension {
}

message PlainTextExtension {
}

message LocalColorTable {
  required bytes colors = 1;
}

message SubBlock {
  required uint32 len = 1;
  required bytes data = 2;
}

message ImageData {
  required uint32 lzw = 1;
  repeated SubBlock subs = 2;
}

message BasicChunk {
  required ImageDescriptor imDescriptor =1;
  required LocalColorTable lct = 2;
  required ImageData img = 3;
  optional GraphicControlExtension gcExt = 4;
}

message PlainTextChunk {
  optional GraphicControlExtension gcExt = 1;
  required PlainTextExtension ptExt = 2;
}

message ApplicationExtension {

}

message CommentExtension {

}

// Stubs
message Header {}
message Trailer {}

message GifProto {
  required Header header = 1;
  required LogicalScreenDescriptor lsd = 2;
  required GlobalColorTable gct = 3;
  repeated ImageChunk chunks = 4;
  required Trailer trailer = 5;
}

package gifProtoFuzzer;